jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererCodeCoverageTracker.kt

```diff
@@ -0,0 +1,143 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.codewhisperer.telemetry
+
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.editor.event.DocumentEvent
+import com.intellij.util.Alarm
+import com.intellij.util.AlarmFactory
+import org.jetbrains.annotations.TestOnly
+import software.aws.toolkits.jetbrains.services.codewhisperer.model.InvocationContext
+import software.aws.toolkits.jetbrains.services.codewhisperer.model.SessionContext
+import software.aws.toolkits.jetbrains.services.codewhisperer.popup.CodeWhispererPopupManager
+import software.aws.toolkits.jetbrains.services.codewhisperer.popup.CodeWhispererUserActionListener
+import software.aws.toolkits.jetbrains.services.codewhisperer.util.CodeWhispererConstants.TOTAL_SECONDS_IN_MINUTE
+import software.aws.toolkits.telemetry.CodewhispererLanguage
+import software.aws.toolkits.telemetry.CodewhispererTelemetry
+import java.time.Duration
+import java.time.Instant
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.math.roundToInt
+
+abstract class CodeWhispererCodeCoverageTracker(
+    private val timeWindowInSec: Long,
+    private val language: CodewhispererLanguage,
+    acceptedTokensSize: AtomicInteger,
+    totalTokensSize: AtomicInteger
+) : Disposable {
+    val percentage: Int
+        get() = if (totalTokensSize.get() != 0) (acceptedTokensSize.get().toDouble() / totalTokensSize.get() * 100).roundToInt() else 0
```

Comment:
Percentage calculation is wrong if we don't use rangeMarker to calculate the correct length of CodeWhisperer suggestion after 5 mins since, CodeWhisperer's suggestion when accepted can very posibilitty be different than the one after 5 mins.

Reply:
yea make sense, added the patch PR